doc/specs/#476 - Package Pinning.md

```diff
@@ -0,0 +1,205 @@
+---
+author: Yao Sun @yao-msft
+created on: 2022-10-12
+last updated: 2022-10-12
+issue id: 476
+---
+
+# Package Pinning
+
+For [#476](https://github.com/microsoft/winget-cli/issues/476)
+
+## Abstract
+
+This spec describes the functionality and high level implementation design of Package Pinning feature.
+
+## Inspiration
+
+This is inspired by functionalities in other package managers, as well as community feedback.
+- Packages may introduce breaking changes that users may not want integrate into their workflow quite yet.
+- Packages may update themselves so that it will be duplicate effort for winget to try to update them.
+- User may want to maintain some of the packages through other channels outside of winget, or prefer one source over others within winget.
+- User may want some of the packages to stay in some major versions but allow minor version changes during upgrade.
+
+## Solution Design
+
+#### Package Pinning types
+To achieve goals listed above, winget will support 3 types of Package Pinning:
+- **Blocking:** The package is blocked from `winget upgrade --all` or `winget upgrade <specific package>`, user has to unblock the package to let winget perform upgrade.
+- **Pinning:** The package is excluded from `winget upgrade --all` but allowed in `winget upgrade <specific package>`, a new argument `--include-pinned` will be introduced to let `winget upgrade --all` to include pinned packages.
+- **Gating:** The package is pinned to specific version(s). For example, if a package is pinned to version `1.2.*`, any version between `1.2.0` to `1.2.<anything>` is considered valid.
+
+To allow user override, `--force` can be used with `winget upgrade <specific package>` to override some of the pinning created above.
+
+#### Package Pinning Configuration Storage
+
+A separate sqlite db (other than the existing tracking catalog) will be created to store the package pinning configurations from user.
+```text
+PackageIdentifier       SourceIdentifier       Version        PinningType
+----------------------------------------------------------------------------
+Microsoft.TestApp       winget                 1.2.*          Gating
+```
+
+**Notes:** For this iteration, winget will only support pinning packages that are locally installed and correlatable with at least one of the remote sources. Winget will record a pinned package by the PackageIdentifier and SourceIdentifier. There can only be one pinning configuration for a specific package. In the future, winget may consider pinning packages from installed packages (upon improving the installed package's PackageIdentifier logic).
+
+## UI/UX Design
+
+#### winget pin commands
+
+A new `winget pin` command with 3 sub-commands will be introduced.
+- Add package pinning configuration:
+  
+  `winget pin add <package> [--version <optional gated version>] [--source <source>] [--force] [--blocking]`
```

Comment:
yes, they'll be same

Reply:
Does that also go for the package query arguments this will use? So, having `--query`, `--id`, `--name`, `--moniker`, `--tag`, `--command`? (I'm already excluding `--manifest` as that is not linked to a source.)

Reply:
yes, any args for finding a package and applicable for search a package to pin.